### PR TITLE
refactors image loading for different dataset formats

### PR DESCRIPTION
**Pull request by grantmerz**
There is a lot in this PR, so here's a list of major changes. It addresses Issues #18 #28 and most of #30 

1. Introduces a ImageReader class.  This is a base class that will take a key from the metadata (e.g. a file name) and use it to read in an image.  Users need to provide the function that takes a key and outputs an image.  The class will also apply scalings to the image data (lupton, z-score) that were previously in the `read_image_hsc`/`read_image_decam` functions in `src/deepdisc/astrodet/astrodet`.  It also lets users add new scalings via add_scaling classmethod
2. Modifies the train_mapper_cls to take an ImageReader as well as a key_mapper function.  The key_mapper function is user-specified and takes a dictionary of metadata as input and outputs the key that will be used to read an image.

Steps 1 and 2 allow for flexibility in the file format of the images.  An image could be saved in multiple files for each band, or a single file containing all bands.  You just need to provide functions that will return an image based on the dictionary data. `test_run_transformers.py` now uses these functions

3. Added testing files for DC2 data.  `test_run_transformers_DC2.py` is pretty much the same as `test_run_transformers.py` except the new functions from steps 1 and 2 and a few cfg parameters are changed to deal with the different data format.  `test_run_transformers_DC2_redshift.py` is also pretty much the same except it includes the redshift estimation network, which the other scripts do not have.
4. Added eval files for DC2 data.  They will spit out true classes and predicted classes. test_eval_model_DC2_redshift will spit out true redshifts and a predicted pdf for each one. 

Run the dc2 test files with `python test_run_transformers_DC2(_redshift).py --data-dir ./tests/deepdisc/test_data/dc2/` 
and `python test_eval_model_DC2(_redshift).py --testfile ./tests/deepdisc/test_data/dc2/single_test.json --nc 1`

### CODE REVIEW

File: src/deepdisc/model/loaders.py [hunk 1]
```diff
@@ -86,69 +55,43 @@ def __call__(self, dataset_dict):
...
 
-        augs = T.AugmentationList(
+        augs = detectron_addons.KRandomAugmentationList(
             [
-                T.CropTransform(
-                    image.shape[1] // 4,
-                    image.shape[0] // 4,
-                    image.shape[1] // 2,
-                    image.shape[0] // 2,
-                )
```
Comment (jeremykubica) [inline src/deepdisc/model/loaders.py:137]: To confirm, it looks like this PR moves the training from using a deterministic cropping to a random one. I think this makes sense, but wanted to check that it was intentional.

File: src/deepdisc/model/loaders.py [hunk 2]
```diff
@@ -176,13 +119,96 @@ def return_train_loader(cfg_loader, **kwargs):
     -------
         a train loader
     """
-    _train_mapper = train_mapper_cls(**kwargs)
-    loader = data.build_detection_train_loader(cfg_loader, mapper=_train_mapper)
+    # _train_mapper = mapper
```
Comment (jeremykubica) [inline src/deepdisc/model/loaders.py:122]: Can we remove this line?

**Review by jeremykubica — COMMENTED** [review]
This looks like a really nice change. I added a few suggestions and, more importantly, clarifying questions to make sure I'm thinking correctly about this.

File: src/deepdisc/data_format/file_io.py [hunk 3]
```diff
@@ -17,3 +20,99 @@ def get_data_from_json(file):
...
+
+
+class ImageReader:
+    """Class that will read images on the fly for the training/testing dataloaders"""
+
+    def raw(im):
+        return im.astype(np.float32)
+
+    def lupton(im, bandlist=[2, 1, 0], stretch=0.5, Q=10, m=0):
+        """
```
[the remark this answers is not in the document]
Comment (grantmerz) [reply src/deepdisc/data_format/file_io.py:32]: Will do

File: src/deepdisc/data_format/file_io.py [hunk 4]
```diff
@@ -17,3 +20,99 @@ def get_data_from_json(file):
...
+        I = np.mean(im, axis=-1)
+        Imean = np.nanmean(I)
+        Isigma = np.nanstd(I)
+
+        for i in range(im.shape[-1]):
+            image[:, :, i] = A * (im[:, :, i] - Imean - m) / Isigma
+
+        return image
+
+    norm_dict = {"raw": raw, "lupton": lupton}
```
[the remark this answers is not in the document]
Comment (grantmerz) [reply src/deepdisc/data_format/file_io.py:76]: Will do

File: src/deepdisc/inference/predictors.py [hunk 5]
```diff
@@ -41,12 +41,7 @@ def get_predictions(dataset_dict, predictor, **kwargs):
...
     """
 
-    # We will want to make this handle other image readers eventually
-    filenames = [
-        dataset_dict["filename_G"],
-        dataset_dict["filename_R"],
-        dataset_dict["filename_I"],
-    ]
-    img = toolkit.read_image_hsc(filenames, **kwargs)
+    key = key_mapper(dataset_dict)
```
[the remark this answers is not in the document]
Comment (grantmerz) [reply src/deepdisc/inference/predictors.py:44]: Good point

File: src/deepdisc/model/loaders.py (hunk 1 above)
Comment (grantmerz) [reply src/deepdisc/model/loaders.py:137]: Yeah, the change should be ok

File: src/deepdisc/model/loaders.py (hunk 2 above)
Comment (grantmerz) [reply src/deepdisc/model/loaders.py:122]: done

File: src/deepdisc/model/loaders.py [hunk 6]
```diff
@@ -176,13 +119,96 @@ def return_train_loader(cfg_loader, **kwargs):
...
+        self.IR = imreader
+        self.km = key_mapper
+
+    def __call__(self, dataset_dict):
+        dataset_dict = copy.deepcopy(dataset_dict)
+        key = self.km(dataset_dict)
+
+        image = self.IR(key)
+
+        augs = T.AugmentationList(
```
[the remark this answers is not in the document]
Comment (grantmerz) [reply src/deepdisc/model/loaders.py:138]: Can eventually move these but I think its ok for no

File: src/deepdisc/data_format/file_io.py [hunk 7]
```diff
@@ -17,3 +20,99 @@ def get_data_from_json(file):
...
+
+    def __init__(self, reader, norm="raw", **scalekwargs):
```
[the remark this answers is not in the document]
Comment (grantmerz) [reply src/deepdisc/data_format/file_io.py:78]: I will fix that

File: src/deepdisc/data_format/file_io.py [hunk 8]
```diff
@@ -17,3 +20,99 @@ def get_data_from_json(file):
     with open(file, "r", encoding="utf-8") as f:
         data = json.load(f)
     return data
+
+
+class ImageReader:
```
[the remark this answers is not in the document]
Comment (grantmerz) [reply src/deepdisc/data_format/file_io.py:25]: Thanks!

**Pull request merged**